llvm/lib/SYCLLowerIR/LowerInvokeSimd.cpp

```diff
@@ -0,0 +1,251 @@
+// - Performs data flow analysis for the first argument to determine which
+//   function link-time constant address it is guaranteed to hold, and replaces
+//   the argument with found "target" function.
+// - Marks target functions with VCStackCall attribute as required by the Intel
+//   GPU backend.
+// TODO:
+// - move VCStackCall markup to Intel GPU-specific part (BE) or design a new
+//   target-neutral attribute for markup.
+// - allow "unknown" function pointers, where actual function's address is not
+//   deducible when BE is ready
+
+#include "llvm/SYCLLowerIR/LowerInvokeSimd.h"
+
+#include "llvm/ADT/SmallPtrSet.h"
+#include "llvm/GenXIntrinsics/GenXMetadata.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/Operator.h"
+#include "llvm/Pass.h"
+
+#define DEBUG_TYPE "LowerInvokeSimd"
+
+using namespace llvm;
+
+namespace {
+class SYCLLowerInvokeSimdLegacyPass : public ModulePass {
+public:
+  static char ID; // Pass identification, replacement for typeid
+  SYCLLowerInvokeSimdLegacyPass() : ModulePass(ID) {
+    initializeSYCLLowerInvokeSimdLegacyPassPass(
+        *PassRegistry::getPassRegistry());
+  }
+
+  // run the LowerESIMD pass on the specified module
+  bool runOnModule(Module &M) override {
+    ModuleAnalysisManager MAM;
+    auto PA = Impl.run(M, MAM);
+    return !PA.areAllPreserved();
+  }
+
+private:
+  SYCLLowerInvokeSimdPass Impl;
+};
+} // namespace
+
+char SYCLLowerInvokeSimdLegacyPass::ID = 0;
+INITIALIZE_PASS(SYCLLowerInvokeSimdLegacyPass, "SYCLLowerInvokeSimd",
+                "Lower SYCL's invoke_simd calls", false, false)
+
+// Public interface to the LowerInvokeSimdPass.
+ModulePass *llvm::createSYCLLowerInvokeSimdPass() {
+  return new SYCLLowerInvokeSimdLegacyPass();
+}
+
+namespace {
+constexpr char INVOKE_SIMD_PREF[] = "_Z21__builtin_invoke_simd";
+
+bool isCast(const Value *V) {
+  int Opc = Operator::getOpcode(V);
+  return (Opc == Instruction::BitCast) || (Opc == Instruction::AddrSpaceCast);
+}
+
+using ValueSetImpl = SmallPtrSetImpl<Value *>;
+using ValueSet = SmallPtrSet<Value *, 4>;
+using ConstValueSetImpl = SmallPtrSetImpl<const Value *>;
+using ConstValueSet = SmallPtrSet<const Value *, 4>;
+
+Value *stripCasts(Value *V) {
+  if (!V->getType()->isPtrOrPtrVectorTy())
+    return V;
+  // Even though we don't look through PHI nodes, we could be called on an
+  // instruction in an unreachable block, which may be on a cycle.
+  ConstValueSet Visited;
+  Visited.insert(V);
+
+  do {
+    if (isCast(V)) {
+      V = cast<Operator>(V)->getOperand(0);
+    }
+    assert(V->getType()->isPtrOrPtrVectorTy() && "Unexpected operand type!");
+  } while (Visited.insert(V).second);
+  return V;
+}
+
+const Value *getSingleUserSkipCasts(const Value *V) {
+  while (isCast(V)) {
+    if (V->getNumUses() != 1) {
+      return nullptr;
+    }
+    V = *(V->user_begin());
+  }
+  return V;
+}
+
+void collectUsesSkipThroughCasts(Value *V, SmallPtrSetImpl<const Use *> &Uses) {
```

Comment:
Minor comment:
I would fix the name of the func a little bit: collectUsersSkippedThroughCasts
It collect users (not uses), and Skip->Skipped